crates/primitives/src/signature/sig.rs

```diff
@@ -0,0 +1,727 @@
+use crate::{
+    hex,
+    signature::{Parity, SignatureError},
+    U256,
+};
+use alloc::vec::Vec;
+use core::str::FromStr;
+
+/// An Ethereum ECDSA signature.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Signature<T> {
+    /// Memoized ecdsa signature (if any)
+    inner: T,
+
+    v: Parity,
+    r: U256,
+    s: U256,
+}
+
+#[cfg(feature = "k256")]
+impl<'a> TryFrom<&'a [u8]> for Signature<k256::ecdsa::Signature> {
+    type Error = SignatureError;
+
+    /// Parses a raw signature which is expected to be 65 bytes long where
+    /// the first 32 bytes is the `r` value, the second 32 bytes the `s` value
+    /// and the final byte is the `v` value in 'Electrum' notation.
+    fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
+        if bytes.len() != 65 {
+            return Err(k256::ecdsa::Error::new().into());
+        }
+        Self::from_bytes_and_parity(&bytes[..64], bytes[64] as u64)
+    }
+}
+
+impl<'a> TryFrom<&'a [u8]> for Signature<()> {
+    type Error = SignatureError;
+
+    fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
+        if bytes.len() != 65 {
+            return Err(SignatureError::FromBytes("expected exactly 65 bytes"));
+        }
+        Self::from_bytes_and_parity(bytes, bytes[64] as u64)
+    }
+}
+
+#[cfg(feature = "k256")]
+impl FromStr for Signature<k256::ecdsa::Signature> {
+    type Err = SignatureError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bytes = hex::decode(s)?;
+        Self::try_from(&bytes[..])
+    }
+}
+
+impl FromStr for Signature<()> {
+    type Err = SignatureError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let bytes = hex::decode(s)?;
+        Self::try_from(&bytes[..])
+    }
+}
+
+impl From<&crate::Signature> for [u8; 65] {
+    #[inline]
+    fn from(value: &crate::Signature) -> [u8; 65] {
+        value.as_bytes()
+    }
+}
+
+impl From<crate::Signature> for [u8; 65] {
+    #[inline]
+    fn from(value: crate::Signature) -> [u8; 65] {
+        value.as_bytes()
+    }
+}
+
+impl From<&crate::Signature> for Vec<u8> {
+    #[inline]
+    fn from(value: &crate::Signature) -> Vec<u8> {
+        value.as_bytes().to_vec()
+    }
+}
+
+impl From<crate::Signature> for Vec<u8> {
```

Comment:
can these impls be generic over T?

Reply:
currently the generic `Signature<T>` type is not reachable by downstream crates. only the aliased parameterized version is. so there's no purpose in making these generic, as they'd only ever be used with the alias